### ISSUE TITLE
Next release

### ISSUE THREAD

**Issue by jtanx** [fontforge/fontforge#3793]
I think the time is right to get the ball rolling for the next release. There's been a fair few bug fixes and user-visible changes, plus there's also been a lot of backend churn.

I'd like to set a target for the 27th-28th, although we can push back if needed.

I'm also aiming for this to be the last release where there is Python 2 support at all, and quite possibly the last release where autotools is used (cmake is coming along quite nicely).

Having said that, if there are major issues with that release, we should always be open to cutting another as some sort of recovery-style release.

Let's tag/link any issues that we think need to be resolved before this next release, @frank-trampe @skef @ctrlcctrlv

**Comment by skef** [comment]
It seems as though Apple hasn't yet announced a core Mac OS X python 3 plan. I know that OS isn't FontForge's biggest platform but until there's a very easy way of getting 3 on there we probably shouldn't drop support for 2.

**Comment by ctrlcctrlv** [comment]
Can we push to August? I want to get my PNG writing in SFDs into the new release. What do you think about that by the way? Also I think I should at least fix #3751 and #3774 needs to be merged before the release, these are known User Decompositions bugs and should be fixed before UD is put into a release

**Comment by skef** [comment]
There are fewer low-hanging fruit now. There are a number of mini-projects that would be valuable but no particular reason I know of to tie one to an upcoming release.

**Comment by jtanx** [comment]
On Python and Mac OS - personally, I'm not too concerned about that. For our own Mac bundles, we're already effectively packaging up our own whole Python distribution, so there's nothing stopping it from switching to Python 3.

Note also that the system Python is being [deprecated](https://www.macobserver.com/analysis/macos-catalina-deprecates-unix-scripting-languages/) without any replacement in sight, so it's likely that you'd need to install Python yourself in the near future anyway.

Homebrew and MacPorts both have Python3 support, and they are the only other two which I know which provide their own packages for FontForge.

@ctrlcctrlv how far into August are we talking? If it's early August, then I'm ok with that. On PNGs in SFDs, the concept sounds ok. My only thoughts are it should be a non-breaking change if possible, and is it possible to just pass-through images (i.e. if I import a jpeg can I just serialise out the original without too much effort).

**Comment by ctrlcctrlv** [comment]
@jtanx Yes, let's say August 1st. I'm okay with that and can get everything in by then.

I have no comments about Mac stuff, you guys can do whatever you want with that, it doesn't affect me and I don't really care (sorry Mac users), I use Arch Linux only.

In re: JPEG. Yes, I do see the benefit. However FontForge throws away the original data. I could probably stop it from doing that somehow...but it's much easier, for now, just to use PNG for any image in FontForge's memory that needs writing to SFD (if `libpng` available). I also plan to make it so we can refer to files _by reference_ which would solve the JPEG problem&hellip;FontForge would re-import then at SFD load time.

**Comment by frank-trampe** [comment]
That timeline works fine for me. The most important thing is that we fix the crashes on i386, and that is done.

The time is certainly approaching to target Python 3 in our builds (and I'll send a note about that to the mailing list), but we cannot drop Python 2 support from the build system. There are definitely a lot of people and workflows using it who will continue to do so.

**Comment by jtanx** [comment]
I don't see why we can't drop it altogether, official Python 2 support is literally ending this year. We've already announced that we're dropping it in the near future too on the mailing list. 

If there really is a workflow that depends on Python 2, there's nothing stopping using an older version of FontForge.

**Comment by khaledhosny** [comment]
AFDKO is dropping Python 2 support, and FontTools as well. These are the two major FOSS font related packages with Python support. People making fonts and requiring Python 2 support are in a bad position already. They can keep using old FontForge along with old Python if they are so inclined.

**Comment by frank-trampe** [comment]
The end of support does not imply an end of use, and I would be surprised if Python 3 overtakes the installed base of Python 2 by the end-of-support date. It doesn't happen to be font stuff, but I still have a lot of code deeply locked into Python 2 due to the lack of Python 3 support in some libraries. Given that we already have pretty good Python-switching capability, it doesn't cost us much to leave this rug underfoot.

**Comment by jtanx** [comment]
I still don't agree with that. As mentioned, if you truly have stuff that's locked to Python 2, then use an older version of FontForge.

> Given that we already have pretty good Python-switching capability, it doesn't cost us much to leave this rug underfoot.

This is not true, I have to implement feature detection with CMake, so there is a cost involved in Python 2 support, both in adding that initial support, and then in ongoing support - which is just a waste of time, because Python 2 is EOL. 

I've already switched Windows builds to Python 3, and I have a plan to switch the Mac bundle to that too. Let's face it, 90% of the time, no one is going to compile FontForge themselves, so there's no good reason to support this into the future.

**Comment by frank-trampe** [comment]
I don't think that it needs to be gracefully supported; it could be as crude as configure-time options for Python version, `PYTHON_CFLAGS`, and `PYTHON_LIBS`. Maybe six lines of code. It may turn out to be a waste of time, and I hope that it is, in fact, but it would be a small one, and I'm happy to do it myself once the main CMake effort is complete.

**Comment by skef** [comment]
@frank-trampe I was going to weigh in more strongly on the "Why not just leave things mostly as they are?" side, but then I realized @jtanx is coming at this primarily from a maintenance angle.

There's quite a bit of python-2-specific code, much of which relates to string handling. Right now (I believe) that code is automatically tested as a side effect of the Mac OS X build. Once that switches over, do we add a special test build for Python 2? Suppose we don't on the theory that python 2 is now an informal thing. Are individual contributors who touch python code now expected to test both versions?  Maintaining parallel python installations isn't *that* complicated, but it definitely raises the personal-IT-expertise bar. So suppose that isn't required either; now the Python 2 support will almost certainly rot. Do we still fix issues if and when filed? 

Personally I'm either/or on this: Either drop support entirely or continue to do an automated build-and-test for Python 2.

**Comment by skef** [comment]
I guess the other thing to add is that there are a number of aspects of the python interface that are weird and make the interface counter-intuitive. (One of these is the extensive use of pseudo-variables for things more appropriately handled by function calls.) So I could see an argument for continuing to support 2 until a point where the interface is overhauled, so that people with older scripts can update them once. 

Obviously any breaking changes to the interface will make a lot of people grumpy ... 

It's also very counter-productive that, for reasons that seem to mostly trace to "packaging", the entirety of fontforge's core python support is implemented in C. It would be far better (again in my view) to have minimal `fontforgeP`, `pointP`, `contourP` classes (or whatever names make more sense) written in C and then have `fontforge`, `point`, `contour` etc. be Python objects that inherit those C objects and have most of the implementation. Yes, this means that the plumbing for FontForge's internal engine needs to be a bit more complicated and to know where the python code "lives". But the current model really limits the extent of python support by making new development so difficult and time-consuming.

**Comment by jtanx** [comment]
I agree with what @skef said.

I strongly dislike the fact that we make things so hard for ourselves by trying to support so many different configuration possibilities. Sure, perhaps supporting Python 2 isn't *too* bad, but it all adds up; it's basically death by a thousand cuts.

Regarding the Python interface, I've long thought we should move towards having a proper `fontforge` package, instead of just dumping `pyd`s directly in the site packages folder, which would make a couple of things packaging wise easier. Regarding directly using CPython, why not use something like Cython? It's a lot of effort though, given what's already in place.

>  It would be far better (again in my view) to have minimal fontforgeP, pointP, contourP classes

I guess that is somewhat alluding to the fact (lib)FontForge has no well defined interface; if we had that then the Python interface would literally just be a binding to that.

**Comment by ctrlcctrlv** [comment]
Is libfontforge having an inconsistent C ABI really such a problem? It makes adding things a lot easier, FontForge is a GUI program, expecting a consistent interface will make things harder.

**Comment by skef** [comment]
> I guess that is somewhat alluding to the fact (lib)FontForge has no well defined interface; if we had that then the Python interface would literally just be a binding to that.

I see this as a separate issue. It would actually be easier given how things currently are to continue to have the "bottom end" of the python interface be implemented using the python C-API. In fact, (the following slightly oversimplifies) you could start out just by giving the current objects new private-ish names, make `fontforge`, `point` and so on empty objects that inherit from their C-API counterparts, and then start "moving things over" if and when it makes sense. 

My objection is just to the extent that things that would be very easy to do in Python and are somewhat onerous to do in C are done in C because of (more or less) plumbing problems.

**Comment by skef** [comment]
@jtanx since you're deep in build infrastructure I encourage you to play around with that model if you get a chance sometime. Basically see if that structure can be set up with the two levels and have the python tests still work (which should confirm they're using the python-level objects because of the naming).

**Comment by jtanx** [comment]
> Is libfontforge having an inconsistent C ABI really such a problem? It makes adding things a lot easier, FontForge is a GUI program, expecting a consistent interface will make things harder.

It's a problem in as much as we currently install (lib)fontforge as something that other packages can link to, even though practically nothing does (I mentioned this before, but only Pdf2HtmlEX does; it's already 
using some fork of FontForge, *and* it's currently abandoned).

If FontForge was designed with something like that in mind, then it would be ok, but it's not. Our headers are not laid out properly. We probably don't even install all required headers now. There's implicit dependencies between all our 'packages'. We don't properly mark which functions should be exported, so we end up needing to export *everything* (this is particularly a problem on Windows). We depend on global variables shared *between* our packages to set preferences. The list goes on...

Which brings me back to the point that adding support for libfontforge as a shared library in CMake is a painful exercise...

**Comment by skef** [comment]
> I've long thought we should move towards having a proper fontforge package, instead of just dumping pyds directly in the site packages folder, which would make a couple of things packaging wise easier. Regarding directly using CPython, why not use something like Cython? It's a lot of effort though, given what's already in place.

On this specific proposal, it could wind up being a false economy if FontForge continues to have an integrated python interpreter, which I assume it will continue to have. You would still have to get things working in both contexts, and given that things currently work in both contexts it's less work to keep the basic model we have. For the reasons I've just laid out nothing about that model forces all of the python development into C: you can just use the C-API as the C/python boundary and put more of the code on the python side by way of inheritance.

**Comment by skef** [comment]
(Some things wouldn't need C-API counterparts. There's (probably?) no reason `psMat` can't be entirely Python as long as one thinks through some basic pickling issues.) `point`s are borderline. A `glyphPen` is mostly glue between other objects and can probably be pure python.

**Comment by skef** [comment]
Hmm -- actually probably just renaming alone wouldn't work because `PyFF_Contour` data (for example) is extracted via a bunch of effectively private interfaces. The code on the C/Python boundary either needs to be consistent about what specifically is passed back and forth, or rewritten to access the python object data via C-python interfaces. The latter would be better for overall coherence (e.g. supporting python-level polymorphism via inheritance).

**Comment by jtanx** [comment]
Somewhat back on topic, mac builds are completely screwed at the moment, they switched to meson to build GTK, but they somehow broke the define that sets `GDK_WINDOWING_QUARTZ` properly.

**Comment by frank-trampe** [comment]
It looks like the fix is merged. Is everything working now?

**Comment by jtanx** [comment]
It was merged upstream, but we use Homebrew for that dependency, so it's kind of dependent of them also applying the fix  - https://github.com/Homebrew/homebrew-core/pull/42217. I'm also seeing if they can just patch/fix Pango too - https://github.com/Homebrew/homebrew-core/pull/42216.

But in the worst case, I can always modify the travis build to manually define `GDK_WINDOWING_QUARTZ`, which should also fix it.

**Comment by frank-trampe** [comment]
Somebody's on the case! We have plenty of time, so I guess we can wait for these to work themselves out.

**Comment by frank-trampe** [comment]
Back to the Python thing, we don't currently change it very often, and the Python 2/3 switching is not the biggest obstacle to changing it very often. Until some significant rewrite, I would like to keep it in place in a semi-usable form. My proposal would be as follows.
* Implement full CMake support for Python 3 (@jtanx).
* Add hotwire support for Python 2 (@frank-trampe).
* Rewrite the Travis rule to build and to test both ways (@frank-trampe).
* Switch all builds to Python 3 (@jtanx and @frank-trampe).
* Announce the Python switch for builds on all media (@frank-trampe and @ctrlcctrlv).
* Spend 2020 reminding people that Python 2 support is deprecated (@frank-trampe).
* Write new rationalized Python bindings in a way that either makes dual-version support elegant or drops Python 2 if that is not possible.

**Comment by skef** [comment]
> Spend 2020 reminding people that Python 2 support is deprecated

This is too conservative. We should start saying now that Python 2 support could potentially last until 2021 but may also be dropped (with a few months warning) if there are significant changes to the python layer. (And dual-version support at this late date seems like a waste of resources.)

And as long as we're "warning" I would favor adding that we may "regularize" the python interface so that most things stay as they are but some of the more counter-intuitive-to-a-python-programmer aspects are (if easy enough) deprecated or (otherwise) eliminated.

**Comment by jtanx** [comment]
This release is done.

**Comment by ctrlcctrlv** [comment]
I tweeted about it https://twitter.com/FontForge/status/1156866405192437760
